README.txt
NAME

  mbuild.lua - Lua-based memoized build system.

DESCRIPTION

  This is an initial implementation of a "memoized build system" in Lua.
  For background on memoized build systems, see [1] and "Related Work" below.

STATUS

  This implementation is an initial proof of concept of creating a
  memoized build system in Lua.  Perhaps it's usable as is, but the
  code will likely need to be expanded in practice.
  
USAGE

  A very basic usage is as follows.  Create a Lua file like this and run it:
  
    local MB = require 'mbuild'
    MB.run('gcc -c t1.c', {'t1.o'}, {'t1.c', 't1.h', 't2.h'})
    MB.run('gcc -c t2.c', {'t2.o'}, {'t2.c', 't2.h'})
    MB.run('gcc -o t t1.c t2.c', {'t'}, {'t1.o', 't2.o'})

  This will compile files "t1.c" and "t2.c" and then link them into a program
  "t".  Inputs and outputs must be listed explicitly (although there are ways
  to implicitly infer them as discussed later below).
  
  If you rerun the program, nothing will happen.  If you edit a dependency and
  rerun the program, the necessary files will be rebuilt.  Checksums will be
  stored in a temporary file called ".deps" (like in fabricate [1]).
  
  You could write wrapper commands that avoid the duplication, like perhaps
  
    c_program('t', {'t1.c', 't2.c'})
  
  You may also wish to put commands in functions:
  
    function program_t()
      MB.run('gcc -c t1.c', {'t1.o'}, {'t1.c', 't1.h', 't2.h'})
      MB.run('gcc -c t2.c', {'t2.o'}, {'t2.c', 't2.h'})
      MB.run('gcc -o t t1.c t2.c', {'t'}, {'t1.o', 't2.o'})
    end
    function program_tt()
      MB.run('gcc -c tt.c', {'tt.o'}, {'tt.c', 'tt.h'})
      MB.run('gcc -o t tt.c', {'t'}, {'tt.o'})
    end
    function all()
      program_t()
      program_tt()
    end
    local name = arg[1] or 'all'
    _G[name]()
  
  Implicit dependencies
  
    Some memoized build systems can invoke "gcc -M" or "gcc -MM"
    to determine dependencies (e.g. [6]).
    An external module is provided for doing this:

      local gcc_deps = require 'mbuild_gcc'.gcc_deps
      MB.run('gcc -c t1.c', {'t1.o'}, {'t1.c', defer=gcc_deps})
      MB.run('gcc -c t2.c', {'t2.c'}, {'t2.c', defer=gcc_deps})
      MB.run('gcc -o t1 t1.c t2.c', {'t1'}, {'t1.o', 't2.o'})
  
    Some memoized builds systems support "strace" (a Linux/*nix command
    not available on Windows) to automatically determine files read and
    written by commands executed. There is a preliminary implementation
    of this via the "mbuild_strace" module:

      local strace = require 'mbuild_strace'
      MB.run('gcc -c t1.c', nil, nil, strace.exec)
      MB.run('gcc -c t2.c', nil, nil, strace.exec)
      MB.run('gcc -o t1 t1.o t2.o', nil, nil, strace.exec)
      
    The strace.exec will build outputs and dependencies based on the
    system calls observed when running the commands.
    
    You can check what dependencies are being inferred by
    examining the ".deps" file that mbuild creates:
    
     ["gcc -c t1.c"]={
      ["/lib/i386-linux-gnu/libc.so.6"]="fae28f0c80586f2b712b191d82a51cbd",
      ["/lib/i386-linux-gnu/libdl.so.2"]="6e2aaea7226f46f38ab50432afd4245d",
      ["t1.c"]="38c4ebbcd1f502ed3f87ae5ec7e6b95b",
      ["t1.h"]="d784fa8b6d98d27699781bd9a7cf19f0",
      . . .
     }

  Change detection
  
    Currently, files are considered "changed" if their MD5 checksums change.
    Alternate possibilities include time-stamp changes.
    
  List utilities
  
    The mbuild_list module provides convenient list manipulation functionality:
    
      local qw = require 'mbuild_list'.qw
      local OBJ = qw'a.o b.o c.o'
      for item in SRC:iter() do
        MB.run('gcc '..item, {item:gsub('%.o$', '.c')}, {item})
      end
      MB.run('gcc -o prog '..OBJ:string(), 'prog', OBJ)
    
EXAMPLES

  examples/build_lua.lua is an example of building Lua 5.2.

DEPENDENCES

  Lua 5.1 or 5.2
  Lua modules:
    file_slurp [8]
    compat_env [9] (required for Lua 5.1 only) 
    md5 [10] (optional if md5sum command available, but recommended for speed)

INSTALLATION / DEPENDENCIES

  Download files and dependencies:
  
    Download all mbuild*.lua files from http://github.com/davidm/lua-mbuild
    wget https://raw.github.com/davidm/lua-file-slurp/master/lua/file_slurp.lua
    wget https://raw.github.com/davidm/lua-compat-env/master/lua/compat_env.lua
         # (omit compat_env in Lua5.2)
    
  Put these in your LUA_PATH.
    
  Either the "md5" module [10] or the system "md5sum" command will be used.
  The later is used if the former is not installed.

RELATED WORK

  Other memoized build systems include [1-5].

DESIGN NOTES

  The core implementation is not aware of particular toolchain, but it does
  support toolchain specific convenience modules like the provided
  mbuild_gcc.lua.  Higher level build rules could be added that abstract
  away the user's toolchain specific code.  This implementation is also
  cross-platform.  (The optional strace support requires
  a Linux/*nix compatible OS.)

REFERENCES

  [1] Fabricate, http://code.google.com/p/fabricate/
  [2] Bill McCloskey's memoize.py.
        Broken link: http://www.eecs.berkeley.edu/~billm/memoize.html .
        Alternate links: http://benno.id.au/memoize.py ;
          http://benno.id.au/blog/2008/06/06/memoize-build-framework
  [3] fbuild - https://github.com/felix-lang/fbuild
        (by Erick Tryzelaar, author of felix-lang)
  [4] mem - http://srp.github.com/mem/getting-started.html
  [5] wonderbuild -
        http://retropaganda.info/~bohan/work/psycle/branches/bohan/wonderbuild/wonderbuild/ ;
        http://psycle.svn.sourceforge.net/viewvc/psycle/branches/bohan/wonderbuild/benchmarks/time.xml
  [6] http://code.google.com/p/fabricate/wiki/HowtoMakeYourOwnRunner
  [7] http://en.wikipedia.org/wiki/Strace
  [8] https://github.com/davidm/lua-file-slurp/
  [9] https://github.com/davidm/lua-compat-env/
  [10] http://www.keplerproject.org/md5/
	
COPYRIGHT

(c) 2012 David Manura.  Licensed under the same terms as Lua 5.1 (MIT license).

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGES.txt
2012-04-28
  - Fixed links to dependent modules.
  - Lua 5.2 compatibility.

2012-01-25 - added initial strace support

2012-01-21 - initial implementation


mbuild_util.lua
--[[
 Various utility functions for mbuild.
 
 (c) 2012 David Manura.  See README for details.
--]]

local M = {}

local FS = require 'file_slurp' -- https://gist.github.com/1325400
local loadfile = (pcall(load,'') and _G or require 'compat_env').loadfile -- https://gist.github.com/1654007

-- Simple table serialize and unserialize from file.
local function dump_string(o, space)
  space = space or ''
  if type(o) == 'string' then
    return ("%q"):format(o)
  elseif type(o) ~= 'table' then
    return tostring(o)
  else
    local ts = {}
    for k,v in pairs(o) do
      ts[#ts+1] = space..'  ['..dump_string(k)..'] = '..dump_string(v, space..'  ')..';\n'
    end
    return '{\n'..table.concat(ts)..space..'}'
  end
end
function M.load_table(filename)
  local f, err = loadfile(filename, 't', {}); if not f then return f, err end
  local t = f()
  return t
end
function M.save_table(filename, t, flags)
  FS.writefile(filename, 'return '..dump_string(t)..'\n', flags)
end

-- MD5 file checksum support.
-- Note: For greater efficiency, use the "md5" module
--   <http://www.keplerproject.org/md5/> if available.
local sumhexa = pcall(require, 'md5') and require 'md5'.sumhexa
if not sumhexa then
  io.stderr:write('Warning: using md5sum command (slow) rather than md5 module.\n')
end
function M.md5_file(filename)
  if sumhexa then
    local data = FS.readfile(filename, 's')
    if not data then return nil end
    return sumhexa(data)
  else
    if not FS.testfile(filename) then return nil end
    local res = FS.readfile('md5sum -b '..filename, 'p')
    local md5 = res:match('[a-f0-9]+') or error("md5sum output bad: filename="..filename)
    return md5
  end
end

function M.concat_lists(t1, t2)
  if #t1 == 0 then return t2 end  -- optimization
  if #t2 == 0 then return t1 end
  local t = {}
  for _,v in ipairs(t1) do t[#t+1] = v end
  for _,v in ipairs(t2) do t[#t+1] = v end
  return t
end

-- Supports __gc metamethods on tables in Lua 5.1.
function M.compat_finalizable(o)
  if _G._VERSION == 'Lua 5.1' then
    M._final = newproxy(true)
    getmetatable(M._final).__gc = function() o:__gc() end
  end
end

return M


mbuild_strace.lua
--[[
*nix strace [1] handler.
 
This is used for automatically determining the files read and written
during a command execution.
The basic technique is described in [2]
 
Warning: this is very preliminary.
 
[1] http://en.wikipedia.org/wiki/Strace
[2] http://code.google.com/p/fabricate/
 
(c) 2012 David Manura.  See README for details.
--]]

local M = {}

local FS = require 'file_slurp'

local DEBUG = false

local tmpfile = 'tmp.txt' -- TODO: generalize location

local function trace_exec(cmd)
  io.stdout:write(': ', cmd, '\n')
  os.remove(tmpfile)
  local cmd2 = 'strace -f -o '..tmpfile..' -e "trace=file" -e "verbose=" '..cmd
  local res = os.execute(cmd2)
  local text = (res == 0) and FS.readfile('tmp.txt', 'T') or nil
  os.remove(tmpfile)  
  return res, text
end

local function hash_to_list(t)
  local tt = {}
  for k in pairs(t) do tt[#tt+1] = k end
  return tt
end

local function parse(text)
  local inputs = {}
  local outputs = {}

  for line in text:gmatch'[^\r\n]+' do
    local filename, flags = line:match'^%d+ open%("([^"]+)", ([^%)]+)'
    if filename then
      local is_write = not flags:match'O_RDONLY'
      ;(is_write and outputs or inputs)[filename] = true
    end
  end
  -- TODO: handle other calls, including chdir (which requires
  -- keeping track of the current working directory for each PID).
  
  -- ignore these files [TODO: generalize]
  inputs['/proc/meminfo'] = nil
  
  return hash_to_list(outputs), hash_to_list(inputs)
end

function M.exec(cmd)
  local res, text = trace_exec(cmd)
  assert(res == 0, 'command failed: '..cmd)
  local outputs, inputs = parse(text)
  if DEBUG then
    for _,v in ipairs(outputs) do print('TRACE_OUT', v) end
    for _,v in ipairs(inputs) do print('TRACE_IN', v) end
  end
  return outputs, inputs
end

return M


examples/build_lua52.lua
#!/usr/bin/env lua
-- Example mbuild for Lua 5.2.0.

local MB = require 'mbuild'
local gcc_deps = require 'mbuild_gcc'.gcc_deps
local qw = require 'mbuild_list'.qw

local CORE_O = qw[[
	lapi.o lcode.o lctype.o ldebug.o ldo.o ldump.o lfunc.o lgc.o llex.o
	lmem.o lobject.o lopcodes.o lparser.o lstate.o lstring.o ltable.o
	ltm.o lundump.o lvm.o lzio.o
]]
local LIB_O = qw[[
	lauxlib.o lbaselib.o lbitlib.o lcorolib.o ldblib.o liolib.o
	lmathlib.o loslib.o lstrlib.o ltablib.o loadlib.o linit.o
]]
local BASE_O = CORE_O + LIB_O + qw'lua.c'

for obj in (BASE_O + qw'lua.o luac.o'):iter() do
  local src = obj:gsub('%.o$', '.c')
  MB.run('gcc -c '..src, {obj}, {src, defer=gcc_deps})
end
MB.run('ar rcu liblua.a '..BASE_O:string(), {'liblua.a'}, BASE_O)
MB.run('gcc -o lua lua.o liblua.a -lm', {'lua'}, {'lua.o', 'liblua.a'})
MB.run('gcc -o luac luac.o liblua.a -lm', {'luac'}, {'luac.o', 'liblua.a'})


mbuild.lua
--[[
mbuild.lua - see README for details.

(c) 2012 David Manura.  See README for details.
--]]

local M = {_TYPE='module', _NAME='mbuild', _VERSION='0.1.20120125'}

local UTIL = require 'mbuild_util'

-- os.execute like function, with partial Lua 5.1/5.2 compatiblity.
local function execute(cmd)
  local result, err, code = os.execute(cmd)
  if type(result) == 'number' then  -- Lua 5.1
    return (result == 0 or nil), 'exit', result
  else
    return result, err, code
  end
end


-- The Builder class.
local Builder = {}; Builder.__index = Builder

function Builder:initialize()
  io.stdout:write'Loading .deps...\n'
  self.deps = UTIL.load_table(".deps", 's') or {}
end

function Builder:finalize()
  if self.deps then
    io.stdout:write'Saving .deps...\n'
    local ok, err = UTIL.save_table('.deps', self.deps)
    if not ok then io.stderr:write(err, '\n') end
    self.deps = nil
  end
end

function Builder:execute(cmd)
  io.stdout:write(': ', cmd, '\n')
  local result, err, code = execute(cmd)
  assert(result, 'failed: cmd='..('%q'):format(cmd)..' [code='..code..']')
end

function Builder:check_deps(cmd, outputs)
   -- Fail if an output lacks a known hash.
  for _,item in ipairs(outputs) do
    if not (self.deps[cmd] or {})[item] then
      return false
    end
  end
  
  -- Fail if a known hash does not match.
  for item, last_hash in pairs(self.deps[cmd] or {}) do
    local hash = UTIL.md5_file(item)
    if hash ~= last_hash then return false end
  end
  
  -- Fail if no info known.
  if not next(self.deps[cmd] or {}) then return false end

  return true -- pass
end

function Builder:update_deps(cmd, outputs, inputs)
  local dep = {}
  if inputs.defer then inputs = inputs:defer() end
  for _,item in ipairs(inputs) do dep[item] = UTIL.md5_file(item) end
  for _,item in ipairs(outputs) do dep[item] = UTIL.md5_file(item) end
  self.deps[cmd] = dep
end

function Builder:run(cmd, outputs, inputs, exec)
  outputs = outputs or {}
  inputs = inputs or {}
  if not self:check_deps(cmd, outputs) then
    local more_outputs, more_inputs = {},{}
    if exec then
      more_outputs, more_inputs = exec(cmd)
    else
      self:execute(cmd)
    end
    local outputs = UTIL.concat_lists(outputs, more_outputs)
    local inputs = UTIL.concat_lists(inputs, more_inputs)
    self:update_deps(cmd, outputs, inputs)
  end
end

function Builder:__gc()
  self:finalize()
end

-- Constructor for Builder objects.
function M.Builder()
  local self = setmetatable({}, Builder)
  UTIL.compat_finalizable(self)
  self:initialize()
  return self
end

-- Optional global builder object for convenience.
M.builder = nil
local function make_global_builder()
  if not M.builder then M.builder = M.Builder() end
  return M.builder
end

-- invokes `run` on global builder object.
function M.run(...)
  local builder = make_global_builder()
  builder:run(...)
end

return M
